perllib/FixMyStreet/Map/WMSBase.pm

```diff
@@ -0,0 +1,274 @@
+        default_zoom   => 7,
+        min_zoom_level => 1,
+        id_offset      => 0,
+    };
+    return $params;
+}
+
+# A hash of parameters used in calculations for map tiles
+sub tile_parameters {
+    my $params = {
+        urls         => [ '' ], # URL of the map tiles, up to the /{z}/{x}/{y} part
+        layer_names  => [ '' ],
+        wms_version => '1.0.0',
+        layer_style  => '',
+        matrix_set   => '',
+        suffix       => '', # appended to tile URLs
+        size         => 256, # pixels
+        dpi          => 96,
+        inches_per_unit => 0, # See OpenLayers.INCHES_PER_UNIT for some options.
+        projection   => 'EPSG:3857', # Passed through to OpenLayers.Projection
+    };
+    return $params;
+}
+
+# This is used to determine which template to render the map with
+sub map_template { 'wms' }
+
+# Reproject a WGS84 lat/lon into an x/y coordinate in this map's CRS.
+# Subclasses will want to override this.
+sub reproject_from_latlon($$$) {
+    my ($self, $lat, $lon) = @_;
+    return (0.0, 0.0);
+}
+
+# Reproject a x/y coordinate from this map's CRS into WGS84 lat/lon
+# Subclasses will want to override this.
+sub reproject_to_latlon($$$) {
+    my ($self, $x, $y) = @_;
+    return (0.0, 0.0);
+}
+
+sub get_res {
+    my ($self, $zoom) = @_;
+
+    my @scales = $self->scales;
+
+    my $res = $scales[$zoom] /
+        ($self->tile_parameters->{inches_per_unit} * $self->tile_parameters->{dpi});
+
+    return $res;
+}
+
+sub map_tiles {
+    my ($self, %params) = @_;
+    my ($left_col, $top_row, $z) = @params{'x_left_tile', 'y_top_tile'};
+    my $tile_url = $self->tile_base_url;
+    my $size = $self->tile_parameters->{size};
+    my $cols = $params{cols};
+    my $rows = $params{rows};
+
+    my @col_offsets = (0.. ($cols-1) );
+    my @row_offsets = (0.. ($rows-1) );
+
+    my $res = $self->get_res($params{zoom});
+    my $size = $res * $size;
+    my ($min_x, $min_y, $max_x, $max_y) = ($left_col, $top_row - $size, $left_col + $size, $top_row);
+
+    return [
+        map {
+            my $row_offset = $_;
+            [
+                map {
+                    my $col_offset = $_;
+                    my $row = $row_offset * $size;
+                    my $col = $col_offset * $size;
+                    my $src = sprintf '%s&bbox=%d,%d,%d,%d',
+                        $tile_url, $min_x + $col, $min_y - $row, $max_x + $col, $max_y - $row;
+                    my $dotted_id = sprintf '%d.%d', ($min_x + $col), ($min_y - $row);
+
+                    # return the data structure for the cell
+                    +{
+                        src => $src,
+                        row_offset => $row_offset,
+                        col_offset => $col_offset,
+                        dotted_id => $dotted_id,
+                        alt => "Map tile $dotted_id",
+                    }
+                }
+                @col_offsets
+            ]
+        }
+        @row_offsets
+    ];
+}
```

Comment:
A lot of this function, and this whole file, is pretty identical/structured same as WMTS base - can we not share these somehow?